pkg/tsdb/azuremonitor/loganalytics/azure-log-analytics-datasource.go

```diff
@@ -95,6 +95,181 @@ func getApiURL(resourceOrWorkspace string, isAppInsightsQuery bool) string {
+	}
+
+	if azureLogAnalyticsTarget.DashboardTime != nil {
+		dashboardTime = *azureLogAnalyticsTarget.DashboardTime
+		if dashboardTime {
+			if azureLogAnalyticsTarget.TimeColumn != nil {
+				timeColumn = *azureLogAnalyticsTarget.TimeColumn
+			} else {
+				// Final fallback to TimeGenerated if no column is provided
+				timeColumn = "TimeGenerated"
+			}
+		}
+	}
+
+	apiURL := getApiURL(resourceOrWorkspace, appInsightsQuery)
+
+	rawQuery, err := macros.KqlInterpolate(query, dsInfo, queryString, "TimeGenerated")
+	if err != nil {
+		return nil, err
+	}
+
+	return &AzureLogAnalyticsQuery{
+		RefID:            query.RefID,
+		ResultFormat:     resultFormat,
+		URL:              apiURL,
+		JSON:             query.JSON,
+		TimeRange:        query.TimeRange,
+		Query:            rawQuery,
+		Resources:        resources,
+		QueryType:        dataquery.AzureQueryType(query.QueryType),
+		AppInsightsQuery: appInsightsQuery,
+		DashboardTime:    dashboardTime,
+		TimeColumn:       timeColumn,
+	}, nil
+}
+
+func buildAppInsightsQuery(ctx context.Context, query backend.DataQuery, dsInfo types.DatasourceInfo, appInsightsRegExp *regexp.Regexp) (*AzureLogAnalyticsQuery, error) {
+	var resultFormat dataquery.ResultFormat
+	dashboardTime := true
+	timeColumn := ""
+	queryJSONModel := types.TracesJSONQuery{}
+	err := json.Unmarshal(query.JSON, &queryJSONModel)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode the Azure Traces query object from JSON: %w", err)
+	}
+
+	azureTracesTarget := queryJSONModel.AzureTraces
+
+	if azureTracesTarget.ResultFormat == nil {
+		resultFormat = types.Table
+	} else {
+		resultFormat = *azureTracesTarget.ResultFormat
+		if resultFormat == "" {
+			resultFormat = types.Table
+		}
+	}
+
+	resources := azureTracesTarget.Resources
+	resourceOrWorkspace := azureTracesTarget.Resources[0]
+	appInsightsQuery := appInsightsRegExp.Match([]byte(resourceOrWorkspace))
+	resourcesMap := make(map[string]bool, 0)
+	if len(resources) > 1 {
+		for _, resource := range resources {
+			resourcesMap[strings.ToLower(resource)] = true
+		}
+		// Remove the base resource as that's where the query is run anyway
+		delete(resourcesMap, strings.ToLower(resourceOrWorkspace))
+	}
+
+	operationId := ""
+	if queryJSONModel.AzureTraces.OperationId != nil && *queryJSONModel.AzureTraces.OperationId != "" {
+		operationId = *queryJSONModel.AzureTraces.OperationId
+		resourcesMap, err = getCorrelationWorkspaces(ctx, resourceOrWorkspace, resourcesMap, dsInfo, operationId)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve correlation resources for operation ID - %s: %s", operationId, err)
+		}
+	}
+
+	queryResources := make([]string, 0)
+	for resource := range resourcesMap {
+		queryResources = append(queryResources, resource)
+	}
+	sort.Strings(queryResources)
+
+	queryString := buildTracesQuery(operationId, nil, queryJSONModel.AzureTraces.TraceTypes, queryJSONModel.AzureTraces.Filters, &resultFormat, queryResources)
```

Comment:
A lot going on here - again I'd be tempted to pull this out into a separate function that returns a struct, and then embed the struct returned by that function into `AzureLogAnalyticsQuery` on L255.

Reply:
Have done this minus the embedded struct aspect as I think it makes things a bit harder to read - wdyt?